src/bruteforce.rs

```diff
@@ -0,0 +1,1129 @@
+
+                                add_root(&approx_forest, left, Range(left_range));
+
+                                // HACK(eddyb) need a more ergonomic SPPF builder/parser API.
+                                parser
+                                    .borrow_mut()
+                                    .with_result_and_remaining(
+                                        Range(right_range),
+                                        Range(full_input.split_at(range.end()).1),
+                                    )
+                                    .forest_add_split(
+                                        rule,
+                                        Node {
+                                            kind: left,
+                                            range: Range(left_range),
+                                        },
+                                    );
+                            }
+
+                            Rule::Or(_) => {
+                                let choice = *i;
+
+                                // HACK(eddyb) need a more ergonomic SPPF builder/parser API.
+                                parser
+                                    .borrow_mut()
+                                    .with_result_and_remaining(
+                                        range,
+                                        Range(full_input.split_at(range.end()).1),
+                                    )
+                                    .forest_add_choice(rule, choice);
+                            }
+
+                            // Only `Concat` and `Or` can be on the stack.
+                            _ => unreachable!(),
+                        }
+                    }
+
+                    // Try to advance this frame.
+                    let mut next = match &approx_forest.possibilities[&(rule, range.start()..)].1 {
+                        SmallSet::One(_) => None,
+                        SmallSet::Many(set) => {
+                            use std::ops::Bound::*;
+                            set.range((Excluded(*i), Unbounded)).next().cloned()
+                        }
+                    };
+                    if let Rule::Concat(_) = cx[rule] {
+                        next = next.filter(|&x| x <= range.len());
+                    }
+
+                    match next {
+                        Some(next) => {
+                            // Advance this frame and keep it around.
+                            *i = next;
+                            continue 'stack;
+                        }
+                        None => {
+                            // Pop this frame and repeat for its parent frame.
+                            valid = *any_valid;
+                            valid_cache.insert((rule, range), valid);
+                            stack.pop();
+                        }
+                    }
+                }
+            }
```

Comment:
So this is the nastiest part of parsing atm, an emulated stack to avoid blowing up the real one.

(Using a queue instead of a stack looks too much like GLL for my own comfort, otherwise I would've probably done it that way)

I'm wondering how I can make this more readable - maybe keep more things explicitly on the stack? Use an enum with variants as an explicit state machine?